framework/src/transfers/MultiAppFieldTransferInterface.C

```diff
@@ -0,0 +1,396 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "MultiAppFieldTransferInterface.h"
+#include "MooseTypes.h"
+#include "FEProblem.h"
+#include "MultiApp.h"
+#include "MooseMesh.h"
+#include "UserObject.h"
+#include "NearestPointIntegralVariablePostprocessor.h"
+
+template <>
+InputParameters
+validParams<MultiAppFieldTransferInterface>()
+{
+  InputParameters params = validParams<MultiAppTransfer>();
+  params.addRequiredParam<std::vector<AuxVariableName>>(
+      "variable", "The auxiliary variable to store the transferred values in.");
+  params.addRequiredParam<std::vector<VariableName>>("source_variable",
+                                                     "The variable to transfer from.");
+  params.addParam<bool>("preserve_transfer",
+                        false,
+                        "Whether or not to conserve the transfered field, "
+                        " if true, the transfered variables will be adjusted "
+                        "according to the pps value");
+
+  std::vector<PostprocessorName> from_postprocessor = {"from_postprocessor"};
+  params.addParam<std::vector<PostprocessorName>>(
+      "from_postprocessor_to_be_preserved",
+      from_postprocessor,
+      "The name of the Postprocessor in the from-app to evaluate an adjusting factor.");
+
+  std::vector<PostprocessorName> to_postprocessor = {"to_postprocessor"};
+  params.addParam<std::vector<PostprocessorName>>(
+      "to_postprocessor_to_be_preserved",
+      to_postprocessor,
+      "The name of the Postprocessor in the to-app to evaluate an adjusting factor.");
+  return params;
+}
+
+MultiAppFieldTransferInterface::MultiAppFieldTransferInterface(const InputParameters & parameters)
+  : MultiAppTransfer(parameters),
+    _from_var_name(getParam<std::vector<VariableName>>("source_variable")),
+    _to_var_name(getParam<std::vector<AuxVariableName>>("variable")),
+    _preserve_transfer(parameters.get<bool>("preserve_transfer")),
+    _from_postprocessor_to_be_preserved(
+        parameters.get<std::vector<PostprocessorName>>("from_postprocessor_to_be_preserved")),
+    _to_postprocessor_to_be_preserved(
+        parameters.get<std::vector<PostprocessorName>>("to_postprocessor_to_be_preserved")),
+    _use_nearestpoint_pps(false)
+{
+  if (_preserve_transfer)
+  {
+    /*
+     * Not sure how important to support multi variables
+     * Let us handle the single variable case only right now if the conservative capability is on
+     */
+    mooseAssert(_to_var_name.size() == 1 && _from_var_name.size() == 1,
```

Comment:
mooseError is preferred.

Reply:
Will not comment on each individual mooseAssert. I think most of them should be mooseError.